jupyterhub/apihandlers/auth.py

```diff
@@ -195,6 +201,30 @@ def needs_oauth_confirm(self, user, oauth_client):
             in self.settings.get('oauth_no_confirm_list', set())
         ):
             return False
+
+        # Check existing authorization
+        existing_tokens = self.db.query(orm.APIToken).filter_by(
+            user_id=user.id,
+            client_id=oauth_client.identifier,
+        )
+        authorized_roles = set()
+        for token in existing_tokens:
+            authorized_roles.update({role.name for role in token.roles})
+
+        if authorized_roles:
+            if set(roles).issubset(authorized_roles):
+                self.log.debug(
+                    f"User {user.name} has already authorized {oauth_client.identifier} for roles {roles}"
+                )
+                # FIXME: uncomment after debugging
+                # it's easier to run manual tests
+                # if we always show the confirmation page
+                # return False
```

Comment:
This is a debug bit to make it easier to run manual tests. It should be put back to `return False` when we are done.